apis/workstations/v1beta1/config_types.go

```diff
@@ -0,0 +1,472 @@
+	//  images](https://cloud.google.com/workstations/docs/custom-container-images).
+	//  If using a private image, the `host.gceInstance.serviceAccount` field
+	//  must be specified in the workstation configuration and must have
+	//  permission to pull the specified image. Otherwise, the image must be
+	//  publicly accessible.
+	Image *string `json:"image,omitempty"`
+
+	// Optional. If set, overrides the default ENTRYPOINT specified by the
+	//  image.
+	Command []string `json:"command,omitempty"`
+
+	// Optional. Arguments passed to the entrypoint.
+	Args []string `json:"args,omitempty"`
+
+	// Optional. Environment variables passed to the container's entrypoint.
+	Env []WorkstationConfig_Container_EnvVar `json:"env,omitempty"`
+
+	// Optional. If set, overrides the default DIR specified by the image.
+	WorkingDir *string `json:"workingDir,omitempty"`
+
+	// Optional. If set, overrides the USER specified in the image with the
+	//  given uid.
+	RunAsUser *int32 `json:"runAsUser,omitempty"`
+}
+
+type WorkstationConfig_Container_EnvVar struct {
+	// Name is the name of the environment variable.
+	Name string `json:"name,omitempty"`
+
+	// Value is the value of the environment variable.
+	Value string `json:"value,omitempty"`
+}
+
+// +kcc:proto=google.cloud.workstations.v1.WorkstationConfig.CustomerEncryptionKey
+type WorkstationConfig_CustomerEncryptionKey struct {
+	// Immutable. A reference to the Google Cloud KMS encryption key. For example,
+	//  `"projects/PROJECT_ID/locations/REGION/keyRings/KEY_RING/cryptoKeys/KEY_NAME"`.
+	//  The key must be in the same region as the workstation configuration.
+	KmsCryptoKeyRef *refs.KMSCryptoKeyRef `json:"kmsCryptoKeyRef,omitempty"`
+
+	// Immutable. A reference to a service account to use with the specified
+	//  KMS key. We recommend that you use a separate service account
+	//  and follow KMS best practices. For more information, see
+	//  [Separation of
+	//  duties](https://cloud.google.com/kms/docs/separation-of-duties) and
+	//  `gcloud kms keys add-iam-policy-binding`
+	//  [`--member`](https://cloud.google.com/sdk/gcloud/reference/kms/keys/add-iam-policy-binding#--member).
+	ServiceAccountRef *refs.IAMServiceAccountRef `json:"serviceAccountRef,omitempty"`
+}
+
+// +kcc:proto=google.cloud.workstations.v1.WorkstationConfig.ReadinessCheck
+type WorkstationConfig_ReadinessCheck struct {
+	// Optional. Path to which the request should be sent.
+	Path *string `json:"path,omitempty"`
+
+	// Optional. Port to which the request should be sent.
+	Port *int32 `json:"port,omitempty"`
+}
+
+// WorkstationConfigSpec defines the desired state of WorkstationConfig
+// +kcc:proto=google.cloud.workstations.v1.WorkstationConfig
+type WorkstationConfigSpec struct {
+	// Parent is a reference to the parent WorkstationCluster for this WorkstationConfig.
+	Parent *WorkstationClusterRef `json:"parentRef"`
+
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="ResourceID field is immutable"
```

Comment:
> the CEL immutability check here will leave the KCC object unfixable in certain situations.

I am not sure that I understand this. What situation would this be a problem? Also, the CEL check is still in the template: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/95b3225e0040b41db93bb05d26bd4d6007dab44b/dev/tools/controllerbuilder/template/apis/types.go#L40

Reply:
Yeah, we should clean up the template as well. Basically this field has been guarded by the resource identity function (check spec.project match externalref.project, spec.location matches  externalref.location, and id matches the  externalRefid). If adding the immutable check in CRD level, certain situations may not work as expected and confuse users. .e.g. user use the default metadata.name (spec.resourceID empty) to create the object, and later on want to set the spec.resourceID, the CEL check fails because nil -> string is considered as a mutable change.